File: Problem_2.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False


class StreamChecker:

    def insert(self, word):
        curr = self.root
        l = len(word)
        for i in range(l-1, -1, -1):
            char = word[i]
            if char not in curr.children:
                curr.children[char] = TrieNode()
            curr = curr.children[char]
        curr.isEnd = True

    def __init__(self, words: List[str]):
        self.root = TrieNode()
        for word in words:
            self.insert(word)
        self.string = []

    def query(self, letter: str) -> bool:
        self.string.append(letter)
        curr = self.root
        # if letter not in curr.children:
        #     return False
        l = len(self.string)
        for i in range(l-1, -1, -1):
            char = self.string[i]
            if char not in curr.children:
                return False
            curr = curr.children[char]
            if curr.isEnd:
                return True

        return False


# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)

File: Problem2.java
// Time Complexity : O(n*l)
// Space Complexity : O(l)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class StreamChecker {
    class TrieNode{
        TrieNode[] children;
        boolean isStart;
        public TrieNode(){
            this.children = new TrieNode[26];
        }
    }
    public void insert(TrieNode root, String word){
        TrieNode curr = root;
        for(int i = word.length()-1; i >=0 ; i--){
            char c = word.charAt(i);
            if(curr.children[c-'a'] == null)
                curr.children[c-'a'] = new TrieNode();
            curr = curr.children[c-'a'];
        }
        curr.isStart = true;
    }
    StringBuilder stream;
    TrieNode root;
    public StreamChecker(String[] words) {
        this.stream = new StringBuilder();
        this.root = new TrieNode();
        for(String word : words){
            insert(root,word);
        }
    }
    
    public boolean query(char letter) {
        stream.append(letter);
        TrieNode curr = root;
        for(int i = stream.length()-1; i >= 0; i--){
            char c = stream.charAt(i);
            if(curr.children[c-'a']==null)
                return false;
            curr = curr.children[c-'a'];
            if(curr.isStart)
                return true;
        }
        return false;
    }
}